### DISCUSSION TITLE
Coroutine was not awaited error when running command.stamp() in async dialect

### DISCUSSION THREAD

**Discussion by EricScottN** [sqlalchemy/alembic#1515] [Usage Questions] [answered]
**Describe the bug**
Even after using `alembic init -t async <script_directory_here>` to generate a env.py file for async dialect, `command.stamp()` will throw an error that `coroutine 'run_async_migrations' was never awaited`. 

**Expected behavior**
Running the following code should create a new version table, and stamp it with the most recent revision as described in [Building an Up to Date Database from Scratch](https://alembic.sqlalchemy.org/en/latest/cookbook.html#building-an-up-to-date-database-from-scratch)
```python
async def setup_db_engine() -> async_sessionmaker | None:
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)
    alembic_cfg = Config("alembic.ini")
    command.stamp(alembic_cfg, "head")
    return async_sessionmaker(engine, expire_on_commit=False)
```

**To Reproduce**
Create a model in sqlalchemy:
```python
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass

class Base(DeclarativeBase, MappedAsDataclass):
    pass

class MyModel(Base):
    __tablename__ = "my_model"

    # Attributes
    id: Mapped[int] = mapped_column(init=False, primary_key=True)
    name: Mapped[str]
```

Generate `env.py` (and other files) with `alembic init -t async <script_directory_here>` and update connection config
```python
config.set_main_option("sqlalchemy.url", <my_connection_string>)
```

Call setup_db_engine() above:
```python
async def main():
    session = await setup_db_engine()
```

**Error**

```
Traceback (most recent call last):
  File "C:\Users\esnor\BuffaloBot\main.py", line 33, in <module>
    asyncio.run(main())
  File "C:\Users\esnor\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\esnor\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\esnor\AppData\Local\Programs\Python\Python312\Lib\asyncio\base_events.py", line 685, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\esnor\BuffaloBot\main.py", line 24, in main
    session = await setup_db_engine()
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\esnor\BuffaloBot\buffalobot\buffalobot.py", line 94, in setup_db_engine
    command.stamp(alembic_cfg, "head")
  File "C:\Users\esnor\BuffaloBot\venv\Lib\site-packages\alembic\command.py", line 690, in stamp
    script.run_env()
  File "C:\Users\esnor\BuffaloBot\venv\Lib\site-packages\alembic\script\base.py", line 583, in run_env
    util.load_python_file(self.dir, "env.py")
  File "C:\Users\esnor\BuffaloBot\venv\Lib\site-packages\alembic\util\pyfiles.py", line 95, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\esnor\BuffaloBot\venv\Lib\site-packages\alembic\util\pyfiles.py", line 113, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "C:\Users\esnor\BuffaloBot\db/migrations\env.py", line 98, in <module>
    run_migrations_online()
  File "C:\Users\esnor\BuffaloBot\db/migrations\env.py", line 92, in run_migrations_online
    asyncio.run(run_async_migrations())
  File "C:\Users\esnor\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 190, in run
    raise RuntimeError(
RuntimeError: asyncio.run() cannot be called from a running event loop
sys:1: RuntimeWarning: coroutine 'run_async_migrations' was never awaited
```

**Versions.**
 - OS: Windows
 - Python: 3.12
 - Alembic: 1.13.2
 - SQLAlchemy: 2.0.29
 - Database: Postgres
 - DBAPI: postgresql+asyncpg

**Additional Context**
Upon examining the `env.py` file the error is coming in this block of generated code:
```python
async def run_async_migrations() -> None:
    """In this scenario we need to create an Engine
    and associate a connection with the context.

    """

    connectable = async_engine_from_config(
        config.get_section(config.config_ini_section, {}),
        prefix="sqlalchemy.",
        poolclass=pool.NullPool,
    )

    async with connectable.connect() as connection:
        await connection.run_sync(do_run_migrations)

    await connectable.dispose()

def run_migrations_online() -> None:
    """Run migrations in 'online' mode."""

    asyncio.run(run_async_migrations())


if context.is_offline_mode():
    run_migrations_offline()
else:
    run_migrations_online()
```

I believe the error here is because `run_async_migrations()` is not being awaited in `asyncio.run(run_async_migrations())` inside the `run_migrations_online()` function. I obviously can't just add `await` as it will be outside of an async function.

Thank you!

**Comment by CaselIT** [comment, marked as answer]
Hi,

You will need to modify your code to conditionally use asyncio.run or the existing loop to run the `run_async_migrations` coroutine.
something like this could work
```py
def run_migrations_online() -> None:
    """Run migrations in 'online' mode."""

    try:
      asyncio.get_running_loop().run_until_complete(run_async_migrations())
      return
    except RuntimeError:
      pass
    asyncio.run(run_async_migrations())

**Reply by CaselIT** [reply]
btw the clue is in the expection: `RuntimeError: asyncio.run() cannot be called from a running event loop`

**Reply by EricScottN** [reply]
Oh my gosh totally missed that part of the trace. Thank you. Runs as expected with what you posted.

Follow up question: Is it worth adding that to the `env.py` that is generated when running `alembic init -t async` so others don't run into that in the future?

**Reply by CaselIT** [reply]
@zzzeek what do you think about adding that change to the template and/or cookbook recipe?

**Reply by zzzeek** [reply]
i certainly dont like the code trying to guess if it's run inside of asyncio already....i like that it raises

im not a fan of this part

```py
async def setup_db_engine() -> async_sessionmaker | None:
    ...
    command.stamp(alembic_cfg, "head")
    ...

```


command.stamp does database IO.  it should be called in a run_sync call.  that would be the part that I'd rather make possible

**Reply by EricScottN** [reply]
> command.stamp does database IO

I checked after running the solution, and there is no alembic version table. TBH, I've never run command.stamp() before this. Is it supposed to create and update that table similar to the normal behavior of `alembic upgrade head`? Isn't that technically the "stamp"?

All the other tables are there from running.
 
```python
async with engine.begin() as conn:
    await conn.run_sync(Base.metadata.create_all)
    ...
```  

Also worth noting I deleted all my migration version files. I know the docs say to "set `down_revision` to `None`" in earliest file and delete the rest, but I figured it could work without any, as well. Maybe that's the reason why.

**Reply by zzzeek** [reply]
i feel like command.stamp() should create the table if it doesnt exist but im not 100% sure.   sorta 95% sure tho

**Reply by zzzeek** [reply]
command.stamp() creates the table so if you have IO going, awaitables being awaited, and importantly transactions being COMMITed, you'll get that table

```
psql (16.3)
Type "help" for help.

test=> select * from alembic_version;
 version_num  
--------------
 185d278c5837
(1 row)

test=> drop table alembic_version;
DROP TABLE
test=> 
\q
[classic@framework alembic:main]$ PYTHONPATH=~/dev/sqlalchemy/lib/ python -m alembic history
5f32c059b464 -> 185d278c5837 (head), rev2
5f32c059b464 -> a815b8836ff0 (head), rev3
<base> -> 5f32c059b464 (branchpoint), rev1
[classic@framework alembic:main]$ PYTHONPATH=~/dev/sqlalchemy/lib/ python -m alembic stamp 185d278c5837
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 185d278c5837
[classic@framework alembic:main]$ psql -U scott test
psql (16.3)
Type "help" for help.

test=> select * from alembic_version;
 version_num  
--------------
 185d278c5837
(1 row)

```

**Reply by fkarg** [reply]
solved it with
```python
def run_migrations_online() -> None:
    """Run migrations in 'online' mode."""

    try:
        # use preexisting event loop if already in async context
        asyncio.ensure_future(run_async_migrations(), loop=asyncio.get_running_loop())
    except RuntimeError:
        asyncio.run(run_async_migrations())
```

**Comment by CaselIT** [comment]
> i certainly dont like the code trying to guess if it's run inside of asyncio already....i like that it raises
> 
> im not a fan of this part
> 
> ```python
> async def setup_db_engine() -> async_sessionmaker | None:
>     ...
>     command.stamp(alembic_cfg, "head")
>     ...
> ```
> 
> command.stamp does database IO. it should be called in a run_sync call. that would be the part that I'd rather make possible

IIRC it's not the first time we get this request. Overall we could have a cookbook example for this.

Doing the proper thing also works but it may not be as clear to read. We can check what it would look like
